write some documentation

tomjnixon opened the pull request:
Preview here:

https://ear.readthedocs.io/en/docs/index.html

Closes #14

AndrewMason commented on doc/BW64.rst:
```diff
@@ -0,0 +1,113 @@
+To read or write a BW64 file, the primary interface is the
+:func:`openBw64Adm` and :func:`openBw64` functions.
+
+To read parsed ADM metadata and samples from a file, use something like:
```
This suggests that the ADM within the file has to be in a format that is (somehow) already parsed.  I think the meaning is
"To read samples, and to read and parse ADM metadata, from a file, use something like:"

tomjnixon replied:
Good point, should be fixed (and generally a bit better).

AndrewMason commented on ear/fileio/utils.py:
```diff
@@ -26,6 +36,18 @@ def openBw64(filename, mode='r', **kwargs):
 
 def openBw64Adm(filename, fix_block_format_durations=False):
+    """Open a BW64 ADM file for reading. This automatically parses the ADM, and
+    adds common definitions and information from the CHNA chunk, accessed
```
The way this is worded suggests that the common definitions are in the CHNA chunk.  Something from my poking around seemed to show me a load of common definitions, perhaps all of them, and suggested to me that it's more along these lines:
"This automatically parses the ADM, adds the information from the CHNA chunk, and adds the common definitions, ..."    But I don't know from where the common defs are added - somewhere they seemed to appear as if by magic.

tomjnixon replied:
Fixed, but in a slightly different way.

The common definitions are added by the XML parsing functions, I just tend to mention it at higher levels too to minimise surprise.

AndrewMason commented on ear/fileio/adm/elements/main_elements.py:
```diff
@@ -195,12 +285,28 @@ def add_encodePackFormat(decode_pack, new_encode_pack):
+    Attributes:
+        audioChannelFormatName (str)
+        type (TypeDefinition): ``typeDefintion`` and/or ``typeLabel``
+        audioBlockFormats (list[AudioBlockFormat])
```
I don't know whether this is the place, but in the docs generated for the AudioChannelFormat class, the entry for https://ear.readthedocs.io/en/docs/ADM_data.html#ear.fileio.adm.elements.AudioChannelFormat does not have a link in place for AudioBlockFormat in  "list[AudioBlockFormat]".  It has a link to the basic "list" type, only.

tomjnixon replied:
I fixed this a while ago (sorry) -- turns out the type was originally called `BlockFormat` not `AudioBlockFormat` for some reason.

AndrewMason commented on ear/fileio/adm/elements/geom.py:
```diff
@@ -27,20 +27,42 @@ def convert_object_position(value):
+    Objects where polar coordinates are used.
+
+    Attributes:
+        azimuth (float)
```
In degrees

tomjnixon replied:
Should be fixed. In general I'd like to rely on correspondence with the ADM though to avoid too much duplication.

AndrewMason commented on ear/fileio/adm/elements/geom.py:
```diff
@@ -27,20 +27,42 @@ def convert_object_position(value):
+
+    Attributes:
+        azimuth (float)
+        elevation (float)
```
In degrees

tomjnixon commented on ear/core/objectbased/conversion.py:
```diff
@@ -110,6 +128,16 @@ def _find_sector(self, az):
         assert False
 
     def point_polar_to_cart(self, az, el, d):
+        """Convert a position from polar to Cartesian.
```
Good point, I added a reference the the specification section from these functions, where the nature of this is explained. There's also a description at the top of the page where this documentation appears.

I don't really want to duplicate a description of this 6 times, as it's quite hard to put in a way which is succinct, accurate and helpful.

tomjnixon commented:
Done, I think
- rebased onto master, removing some commits I added accidentally
- squashed lots of commits together so the history makes a bit more sense
  - generally commits are small fixes found, documentation comment tidying/improvements, or adding the sphinx doc structure
  - I added a copy of sphinxarg with a few changes to support the latest sphinx versions. It might be good to upstream this, but it seems a bit abandoned.
